alibi_detect/utils/pytorch/prediction.py

```diff
@@ -1,12 +1,14 @@
+from functools import partial
 import numpy as np
 import torch
 import torch.nn as nn
 from typing import Callable, Union
+from alibi_detect.utils.prediction import tokenize_transformer
 
 
-def predict_batch(x: Union[np.ndarray, torch.Tensor], model: Union[nn.Module, nn.Sequential],
-                  device: torch.device = None, batch_size: int = int(1e10),
-                  dtype: Union[np.float32, torch.dtype] = np.float32) -> Union[np.ndarray, torch.Tensor]:
+def predict_batch(x: Union[list, np.ndarray, torch.Tensor], model: Union[Callable, nn.Module, nn.Sequential],
+                  device: torch.device = None, batch_size: int = int(1e10), preprocess_fn: Callable = None,
+                  dtype: Union[np.float32, torch.dtype] = np.dtype) -> Union[np.ndarray, torch.Tensor, tuple]:
```

Comment:
I don't think these `dtype` annotations are doing what you mean/want...

Reply:
Unsure if even `dtype: Union[np.dtype, torch.dtype] = np.float32` would be correct. Thoughts?

Reply:
Unlikely, if in doubt, check with `isinstance`. We had this conversation previously, but I forgot what the conclusion was...

Reply:
I.e. `np.float32` is not an instance of `np.dtype`, there is something subtle going on there.

Reply:
Probably needs a separate issue to look at how to type this properly (and if it's even possible).

Reply:
I can't seem to find a nice way to cover all numpy types without explicitly listing them in `dtype`. It's a little bit confusing because `isinstance(x.dtype, np.dtype)=True` but `type(np.float32)=type` and `isinstance(np.float32, np.dtype)=False` but e.g. `isinstance(torch.float32, torch.dtype)=True`.

Reply:
Both `np.dtype` and `torch.dtype` are of type `type` (because they are classes), so `dtype: type = np.float32` is valid, but not very useful as it doesn't narrow it down. E.g. see https://github.com/SeldonIO/alibi-detect/blob/5be1ed9f240112089b492d3f98905b837b3298cf/alibi_detect/utils/prediction.py#L27

Reply:
Yes, even if strictly not correct, at least in this PR it should be clear that you either pass a numpy or torch type. So leave as is and open issue?

Reply:
Yea I guess so

Reply:
#281